myia/vm.py

```diff
@@ -0,0 +1,193 @@
+"""Debug/Testing Virtual Machine.
+
+This VM will directly execute a graph so it should be suitable for
+testing or debugging.  Don't expect stellar performance from this
+implementation.
+"""
+from typing import List, Dict, Any
+
+from myia.anf_ir import Graph, ANFNode, Constant, Parameter, Apply
+from myia.primops import Primitive, Add, If, Return
+
+
+class VM:
+    """
+    Virtual Machine interface.
+
+    Attributes:
+        env: evaluation environement (currently unused)
+
+    """
+
+    def __init__(self, env) -> None:
+        """Construct a VM."""
+        self.env = env
+
+    def evaluate(self, graph: Graph, args: List[Any]):
+        """
+        Run a graph.
+
+        This will evaluate the passed-in graph and return the
+        resulting value.
+
+        """
+        frame = VMFrame(graph, args, self.env)
+        return frame.evaluate()
+
+
+class VMFrame:
+    """
+    An execution frame.
+
+    This is the main class for the VM.  It does all the bookkeeping to
+    handle the execution state.
+
+    Attributes:
+        graph: The currently executing graph.  This can change due to
+            a tail call.
+        args: Argument array for the current call.
+        values: Mapping of every node in the graph to its computed
+            value.  Currently all the values are always kept live.
+        todo: List of node that have to be visited.  Evaluation starts
+            from the end of the list.
+
+    """
+
+    class Continue(Exception):
+        """Continue step."""
+
+    class Return(Exception):
+        """End step."""
+
+    def __init__(self, graph: Graph, args: List[Any], env) -> None:
+        """Build a frame."""
+        self.env = env
+        self.values: Dict[ANFNode, Any] = dict()
+        self.reset(graph, args)
+        self.calling: VMFrame = None
+
+    def reset(self, graph: Graph, args: List[Any]) -> None:
+        """
+        Set up a call.
+
+        This is used while the frame is executing a graph to implement
+        tail calls.
+        """
+        self.graph = graph
+        self.args = args
+        self.todo: List[ANFNode] = [self.graph.return_]
+
+    def evaluate(self):
+        """Run the graph to completion and return the result."""
+        while True:
+            try:
+                self.advance()
+            except StopIteration:
+                return self.values[self.graph.return_]
+
+    def tail(self):
+        """Are we in a tail call?."""
+        return self.todo[-2] is self.graph.return_
+
+    def done(self):
+        """Are we done?."""
+        return self.graph.return_ in self.values
```

Comment:
Not with the current design, since we put the return node as the last one on the todo list.

If you have an example of a graph that could need to evaluate some other nodes after the return, please show me an example.

Reply:
I'm thinking of something like this:

```python
def f(x):
    a = x * x
    def g():
        return a
    return g

def h():
    g = f(9)
    return g() + g()
```

There is no direct path from `f.return_` to the node for `a`: it is the graph for `g` that has a pointer to `a` which belongs to `f`. Nonetheless, it needs to be evaluated in `f`'s environment, at most once per `g` closure, and should reuse anything computed when evaluating `f`. However, both calls to `g` happen after `f` is done evaluating, so unless we walk `g` to find `a` and add it to the todo list immediately (i.e. closure conversion), `a` is going to be found after the frame for `f` exited.

Reply:
Oh yeah, I see the problem.  It will attempt to recompute `a`, but the parameters will no longer be available so it will fail or produce bad results.